python/cudf/cudf/core/join/join.py

```diff
@@ -0,0 +1,454 @@
+                "{!r} merge not supported yet".format(how)
+            )
+
+        # Passing 'on' with 'left_on' or 'right_on' is ambiguous
+        if on and (left_on or right_on):
+            raise ValueError(
+                'Can only pass argument "on" OR "left_on" '
+                'and "right_on", not a combination of both.'
+            )
+
+        # Can't merge on unnamed Series
+        if (isinstance(lhs, cudf.Series) and not lhs.name) or (
+            isinstance(rhs, cudf.Series) and not rhs.name
+        ):
+            raise ValueError("Can not merge on unnamed Series")
+
+        # Keys need to be in their corresponding operands
+        if on:
+            on_keys = [on] if not isinstance(on, list) else on
+            for key in on_keys:
+                if not (key in lhs._data.keys() and key in rhs._data.keys()):
+                    raise KeyError("on key {} not in both operands".format(on))
+        elif left_on and right_on:
+            left_on_keys = (
+                [left_on] if not isinstance(left_on, list) else left_on
+            )
+            right_on_keys = (
+                [right_on] if not isinstance(right_on, list) else right_on
+            )
+
+            for key in left_on_keys:
+                if key not in lhs._data.keys():
+                    raise KeyError('Key "{}" not in left operand'.format(key))
+            for key in right_on_keys:
+                if key not in rhs._data.keys():
+                    raise KeyError('Key "{}" not in right operand'.format(key))
+
+        # Require same total number of columns to join on in both operands
+        len_left_on = 0
+        len_right_on = 0
+        if left_on:
+            len_left_on += (
+                len(left_on) if pd.api.types.is_list_like(left_on) else 1
+            )
+        if right_on:
+            len_right_on += (
+                len(right_on) if pd.api.types.is_list_like(right_on) else 1
+            )
+        if not (len_left_on + left_index * lhs._num_indices) == (
+            len_right_on + right_index * rhs._num_indices
+        ):
+            raise ValueError(
+                "Merge operands must have same number of join key columns"
+            )
+
+        # If nothing specified, must have common cols to use implicitly
+        same_named_columns = set(lhs._data.keys()) & set(rhs._data.keys())
+        if (
+            not (left_index or right_index)
+            and not (left_on or right_on)
+            and len(same_named_columns) == 0
+        ):
+            raise ValueError("No common columns to perform merge on")
+
+        if suffixes:
+            lsuffix, rsuffix = suffixes
+        for name in same_named_columns:
+            if name == left_on == right_on:
+                continue
+            elif left_on and right_on:
+                if (name in left_on and name in right_on) and (
+                    left_on.index(name) == right_on.index(name)
+                ):
```

Comment:
I'm not sure that works here because we're specifically looking for columns that aren't part of the specified join keys. If we just loop through `left_on` and `right_on` we don't have a chance of detecting those.